### ISSUE TITLE
classes appear duplicated in completions when exported by multiple imported libraries

### ISSUE THREAD

**Issue by pq** [dart-lang/sdk#40216]
**TL;DR:** classes appear duplicated when they are exported by multiple imported libraries

**Example:** `FadeInImage`

![image](https://user-images.githubusercontent.com/67586/72652144-672b5180-393a-11ea-9de8-39dc5cb02dec.png)

Reproducible in a file w/ the following imports:

```dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
```

(fyi @keertip, this could be the same issue you encountered internally.)

**Comment by bwilkerson** [comment]
There was an earlier discussion about this (I don't remember where at the moment) in which several alternatives were discussed. I think it was an issue about special casing this behavior for flutter and we were trying to decide whether there was a general algorithm that we could use to solve the problem.

**Comment by pq** [comment]
@DanTup  can confirm but a quick glance at the VSCode implementation suggestions they're doing de-duping based on an identifier consed up like so:

```ts
const fullItemKey = `${suggestion.label}/${suggestion.element.kind}/${suggestion.declaringLibraryUri}`;
```

The set of `includedItems` is initialized here:

https://github.com/Dart-Code/Dart-Code/blob/d4e98d2ca2636be5da7334760d73face12414e70/src/extension/providers/dart_completion_item_provider.ts#L251-L253

**Comment by DanTup** [comment]
Yeah, we use the declaring library URI to avoid showing the same element twice even if it was imported via a different library. My understanding is that this can't be done in the server because these completions are being built client-side (from AvailableSuggestionSets).

I also had a fix at https://github.com/Dart-Code/Dart-Code/issues/1958 / https://github.com/Dart-Code/Dart-Code/commit/ce0087aef29464c5bb324261d1c44aba245dc603 which is to avoid including dupes where there are already multiple imports that bring that item in.

**Comment by bwilkerson** [comment]
It would be possible to do this on the server, without modification of the protocols, though it would take a bit of effort to implement it. The real question is whether we ought to. Or perhaps the question is _when_ we ought to.

For example, given two public libraries L1 and L2 that both export a class C from a private library, and a compilation unit in which neither L1 nor L2 are currently imported, should we offer the user a choice of whether to import C through L1 or L2?

Of course, if one of the libraries is already imported, then we should have a single suggestion for C because we don't need to add an import in that case.

**Comment by DanTup** [comment]
> For example, given two public libraries L1 and L2 that both export a class C from a private library, and a compilation unit in which neither L1 nor L2 are currently imported, should we offer the user a choice of whether to import C through L1 or L2?

In VS Code today, you get both of these in the completion (and the completion detail says "Auto-import from xyz" so you can pick which one).

The de-duping code mentioned above (in VS Code) is for the second case, to ensure that when we add all the symbols for the included suggestion sets, we don't add the same declared class twice because it's in two of them (I believe this is the issue @pq is seeing based on the imports at the bottom of the original comments).

I thought this wouldn't be solvable on the server, because IIRC the server doesn't tell us each symbol to include, it just says things like "all classes and constructors from library x" and "all classes and constructors from library y", which is why I filtered out the dupes client-side (and suspected IntelliJ would need to do the same).

**Comment by pq** [comment]
See also: https://github.com/Dart-Code/Dart-Code/issues/1835.

**Comment by DanTup** [comment]
Good spot - that seems like a server bug I never raised here (though from what I can tell, it only affected enum *values* but your screenshot here isn't?)

**Comment by FMorschel** [comment]
I can't reproduce this anymore. Is there any case where this still happens, or can we close this?

**Comment by keertip** [comment]
It is likely this was fixed with the rewrite of the completion engine. Marking as closed, as it is not reproducible.
